app/src/main/java/de/westnordost/streetcomplete/quests/building_entrance_reference/AddEntranceReference.kt

```diff
@@ -0,0 +1,95 @@
+    """.toElementFilterExpression() }
+
+    override val changesetComment = "Add entrance identification"
+    override val wikiLink = "Key:ref"
+    override val icon = R.drawable.ic_quest_power // TODO!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    override val questTypeAchievements = listOf(CITIZEN, BLIND)
+    override val enabledInCountries = NoCountriesExcept("BG", "PL", "RU", "UA", "BY", "EE")
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_entrance_reference
+
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter { it.tags.containsKey("entrance") }.asSequence()
+
+    override fun isApplicableTo(element: Element): Boolean? =
+        if (element !is Node || (element.tags["entrance"] != "main" && element.tags["entrance"] != "staircase") ) false else null
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        // note: it does not support multipolygon buildings
+        val buildingWays = mapData.ways.asSequence()
+            .filter { buildingFilter.matches(it) }
+        val returned = mutableListOf<Node>()
+
+        for (way in buildingWays) {
+            val entrancesForApartments = mutableListOf<Node>()
+            way.nodeIds.forEach { nodeId ->
+                val node = mapData.getNode(nodeId)
+                if(node != null) {
+                    if(node.tags["entrance"] == "main" || node.tags["entrance"] == "staircase") {
+                        entrancesForApartments.add(node)
+                    }
+                }
+            }
+            if(entrancesForApartments.count() >= 2) {
+                for (entrance in entrancesForApartments) {
+                    if(entrance.tags["noref"] == "yes") {
+                        continue
+                    }
+                    if(entrance.tags["ref:signed"] == "no") {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("addr:flats")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("name")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("ref")) {
+                        continue
+                    }
+                    returned.add(entrance)
+                }
+            }
+        }
+        return returned
+    }
+
+    override fun createForm() = AddEntranceReferenceForm()
+
+    override fun applyAnswerTo(answer: EntranceAnswer, tags: Tags, timestampEdited: Long) {
+        when (answer) {
+            is FlatRange -> tags["addr:flats"] = answer.flatRange
+            is ReferenceCode -> tags["ref"] = answer.referenceCode
+            is ReferenceCodeAndFlatRange -> {
+                tags["addr:flats"] = answer.flatRange
+                tags["ref"] = answer.referenceCode
+            }
+            Unsigned -> tags["ref:signed"] = "no"
```

Comment:
Assignments should be in their own block:

```suggestion
            is FlatRange -> {
                tags["addr:flats"] = answer.flatRange
            }
            is ReferenceCode -> {
                tags["ref"] = answer.referenceCode
            }
            is ReferenceCodeAndFlatRange -> {
                tags["addr:flats"] = answer.flatRange
                tags["ref"] = answer.referenceCode
            }
            Unsigned -> {
                tags["ref:signed"] = "no"
            }
```